File: Backend/src/main/java/com/woovi/crudbank/account/application/AccountService.java
package com.woovi.crudbank.account.application;

import com.woovi.crudbank.account.api.AccountView;
import com.woovi.crudbank.account.domain.AccountDocument;
import com.woovi.crudbank.account.domain.AccountStatus;
import com.woovi.crudbank.account.domain.CurrencyCode;
import com.woovi.crudbank.account.infrastructure.AccountRepository;
import com.woovi.crudbank.shared.error.DomainException;
import com.woovi.crudbank.shared.relay.CursorCodec;
import com.woovi.crudbank.shared.relay.GlobalIdCodec;
import com.woovi.crudbank.shared.util.AccountFieldNormalizer;
import graphql.relay.Connection;
import graphql.relay.ConnectionCursor;
import graphql.relay.DefaultConnection;
import graphql.relay.DefaultConnectionCursor;
import graphql.relay.DefaultEdge;
import graphql.relay.DefaultPageInfo;
import graphql.relay.Edge;
import com.woovi.crudbank.shared.util.MoneyValidator;
import org.springframework.data.domain.Sort;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;

@Service
public class AccountService {

    private static final Sort DEFAULT_SORT = Sort.by(
        Sort.Order.desc("createdAt"),
        Sort.Order.desc("id")
    );

    private final AccountRepository accountRepository;
    private final MongoTemplate mongoTemplate;
    private final GlobalIdCodec globalIdCodec;
    private final CursorCodec cursorCodec;

    public AccountService(
        AccountRepository accountRepository,
        MongoTemplate mongoTemplate,
        GlobalIdCodec globalIdCodec,
        CursorCodec cursorCodec
    ) {
        this.accountRepository = accountRepository;
        this.mongoTemplate = mongoTemplate;
        this.globalIdCodec = globalIdCodec;
        this.cursorCodec = cursorCodec;
    }

    public AccountView createAccount(String ownerName, String document, String branch, String number, BigDecimal initialBalance) {
        String normalizedOwnerName = AccountFieldNormalizer.normalizeOwnerName(ownerName);
        String normalizedDocument = AccountFieldNormalizer.normalizeDocument(document);
        String normalizedBranch = AccountFieldNormalizer.normalizeBranch(branch);
        String normalizedNumber = AccountFieldNormalizer.normalizeAccountNumber(number);

        BigDecimal normalizedBalance = initialBalance == null
            ? BigDecimal.ZERO.setScale(2)
            : MoneyValidator.validateNonNegativeAmount(initialBalance, "initialBalance");

        AccountDocument account = new AccountDocument();
        account.setOwnerName(normalizedOwnerName);
        account.setDocument(normalizedDocument);
        account.setBranch(normalizedBranch);
        account.setNumber(normalizedNumber);
        account.setCurrency(CurrencyCode.BRL);
        account.setCurrentBalance(normalizedBalance);
        account.setStatus(AccountStatus.ACTIVE);
        account.setCreatedAt(Instant.now());
        account.setUpdatedAt(account.getCreatedAt());

        try {
            return toView(accountRepository.save(account));
        } catch (DuplicateKeyException ex) {
            throw DomainException.conflict("An account with same branch and number already exists");
        }
    }

    public AccountView updateAccount(String globalId, String ownerName, String document) {
        AccountDocument account = findByGlobalId(globalId);
        ensureAccountActive(account);

        if (StringUtils.hasText(ownerName)) {
            account.setOwnerName(AccountFieldNormalizer.normalizeOwnerName(ownerName));
        }
        if (StringUtils.hasText(document)) {
            account.setDocument(AccountFieldNormalizer.normalizeDocument(document));
        }

        account.setUpdatedAt(Instant.now());
        return toView(accountRepository.save(account));
    }

    public AccountView deactivateAccount(String globalId) {
        AccountDocument account = findByGlobalId(globalId);
        account.setStatus(AccountStatus.INACTIVE);
        account.setUpdatedAt(Instant.now());
        return toView(accountRepository.save(account));
    }

    public AccountDocument findRawByGlobalId(String globalId) {
        GlobalIdCodec.DecodedGlobalId decoded = globalIdCodec.decode(globalId);
        if (!"Account".equals(decoded.type())) {
            throw DomainException.validation("Expected Account global id");
        }
        return accountRepository.findById(decoded.id())
            .orElseThrow(() -> DomainException.notFound("Account not found"));
    }

    public String decodeAccountId(String globalId) {
        return findRawByGlobalId(globalId).getId();
    }

    public AccountDocument findById(String accountId) {
        return accountRepository.findById(accountId)
            .orElseThrow(() -> DomainException.notFound("Account not found"));
    }

    public AccountView getByGlobalId(String globalId) {
        return toView(findByGlobalId(globalId));
    }

    public Connection<AccountView> listAccounts(Integer first, String afterCursor, String statusFilter) {
        int size = first == null ? 20 : first;
        if (size <= 0 || size > 100) {
            throw DomainException.validation("first must be between 1 and 100");
        }

        Query query = new Query();
        query.with(DEFAULT_SORT);
        query.limit(size + 1);

        if (StringUtils.hasText(statusFilter)) {
            query.addCriteria(Criteria.where("status").is(parseStatus(statusFilter)));
        }

        if (StringUtils.hasText(afterCursor)) {
            CursorCodec.DecodedCursor decodedCursor = cursorCodec.decode(afterCursor);
            Criteria olderDate = Criteria.where("createdAt").lt(decodedCursor.createdAt());
            Criteria sameDateOlderId = new Criteria().andOperator(
                Criteria.where("createdAt").is(decodedCursor.createdAt()),
                Criteria.where("id").lt(decodedCursor.id())
            );
            query.addCriteria(new Criteria().orOperator(olderDate, sameDateOlderId));
        }

        List<AccountDocument> documents = mongoTemplate.find(query, AccountDocument.class);
        boolean hasNextPage = documents.size() > size;
        if (hasNextPage) {
            documents = documents.subList(0, size);
        }

        List<Edge<AccountView>> edges = new ArrayList<>();
        for (AccountDocument document : documents) {
            String cursor = cursorCodec.encode(document.getCreatedAt(), document.getId());
            edges.add(new DefaultEdge<>(toView(document), new DefaultConnectionCursor(cursor)));
        }

        ConnectionCursor startCursor = edges.isEmpty() ? null : edges.getFirst().getCursor();
        ConnectionCursor endCursor = edges.isEmpty() ? null : edges.getLast().getCursor();
        DefaultPageInfo pageInfo = new DefaultPageInfo(startCursor, endCursor, afterCursor != null, hasNextPage);

        return new DefaultConnection<>(edges, pageInfo);
    }

    public AccountView toView(AccountDocument account) {
        return new AccountView(
            globalIdCodec.encode("Account", account.getId()),
            account.getOwnerName(),
            account.getDocument(),
            account.getBranch(),
            account.getNumber(),
            account.getCurrency().name(),
            account.getCurrentBalance(),
            account.getStatus().name(),
            account.getVersion() == null ? null : String.valueOf(account.getVersion()),
            toOffsetDateTime(account.getCreatedAt()),
            toOffsetDateTime(account.getUpdatedAt())
        );
    }

    private static OffsetDateTime toOffsetDateTime(Instant instant) {
        return instant == null ? null : instant.atOffset(ZoneOffset.UTC);
    }

    public void ensureAccountActive(AccountDocument account) {
        if (account.getStatus() != AccountStatus.ACTIVE) {
            throw DomainException.accountInactive("Account is inactive");
        }
    }

    private AccountStatus parseStatus(String status) {
        try {
            return AccountStatus.valueOf(status);
        } catch (IllegalArgumentException ex) {
            throw DomainException.validation("Invalid account status");
        }
    }

    private AccountDocument findByGlobalId(String globalId) {
        return findRawByGlobalId(globalId);
    }
}

File: backend/src/modules/account/accountService.ts
import Decimal from 'decimal.js';
import {
  ObjectId,
  type Collection,
  type Db,
  type Document,
  type Filter,
  type WithId,
} from 'mongodb';
import { DomainError } from '../../shared/errors/DomainError';
import { decodeCursor, encodeCursor } from '../../shared/relay/cursorCodec';
import { decodeGlobalId, encodeGlobalId } from '../../shared/relay/globalIdCodec';
import {
  normalizeAccountNumber,
  normalizeBranch,
  normalizeDocument,
  normalizeOwnerName,
} from '../../shared/validators/accountFieldNormalizer';
import { validateNonNegativeAmount } from '../../shared/validators/moneyValidator';
import { decimalToDecimal128, decimalToString, toDecimal } from '../../shared/utils/decimal';

export type AccountStatus = 'ACTIVE' | 'INACTIVE';

type CurrencyCode = 'BRL';

export interface AccountDocument extends Document {
  _id?: ObjectId;
  ownerName: string;
  document: string;
  branch: string;
  number: string;
  currency: CurrencyCode;
  currentBalance: unknown;
  status: AccountStatus;
  version?: number;
  createdAt: Date;
  updatedAt: Date;
}

export interface AccountView {
  __typename: 'Account';
  id: string;
  ownerName: string;
  document: string;
  branch: string;
  number: string;
  currency: string;
  currentBalance: string;
  status: string;
  version: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface PageInfoView {
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor: string | null;
  endCursor: string | null;
}

export interface ConnectionView<TNode> {
  edges: Array<{ cursor: string; node: TNode }>;
  pageInfo: PageInfoView;
}

function accountCollection(db: Db): Collection<AccountDocument> {
  return db.collection<AccountDocument>('accounts');
}

function toObjectId(id: string, errorMessage = 'Account not found'): ObjectId {
  if (!ObjectId.isValid(id)) {
    throw DomainError.notFound(errorMessage);
  }
  return new ObjectId(id);
}

export class AccountService {
  private readonly db: Db;

  constructor(db: Db) {
    this.db = db;
  }

  async createAccount(
    ownerName: string,
    document: string,
    branch: string,
    number: string,
    initialBalance?: unknown,
  ): Promise<AccountView> {
    const normalizedOwnerName = normalizeOwnerName(ownerName);
    const normalizedDocument = normalizeDocument(document);
    const normalizedBranch = normalizeBranch(branch);
    const normalizedNumber = normalizeAccountNumber(number);

    const normalizedBalance =
      initialBalance === null || initialBalance === undefined
        ? new Decimal(0).toDecimalPlaces(2)
        : validateNonNegativeAmount(initialBalance, 'initialBalance');

    const now = new Date();
    const payload: AccountDocument = {
      ownerName: normalizedOwnerName,
      document: normalizedDocument,
      branch: normalizedBranch,
      number: normalizedNumber,
      currency: 'BRL',
      currentBalance: decimalToDecimal128(normalizedBalance),
      status: 'ACTIVE',
      version: 0,
      createdAt: now,
      updatedAt: now,
    };

    try {
      const result = await accountCollection(this.db).insertOne(payload);
      const created = await accountCollection(this.db).findOne({ _id: result.insertedId });
      if (!created) {
        throw DomainError.badRequest('Unable to create account');
      }
      return this.toView(created);
    } catch (error) {
      if (isDuplicateKeyError(error)) {
        throw DomainError.conflict('An account with same branch and number already exists');
      }
      throw error;
    }
  }

  async updateAccount(globalId: string, ownerName?: string | null, document?: string | null): Promise<AccountView> {
    const account = await this.findRawByGlobalId(globalId);
    this.ensureAccountActive(account);

    const update: Partial<AccountDocument> = {
      updatedAt: new Date(),
      version: (account.version ?? 0) + 1,
    };

    if (typeof ownerName === 'string' && ownerName.trim().length > 0) {
      update.ownerName = normalizeOwnerName(ownerName);
    }

    if (typeof document === 'string' && document.trim().length > 0) {
      update.document = normalizeDocument(document);
    }

    await accountCollection(this.db).updateOne({ _id: account._id }, { $set: update });
    const updated = await this.findById(account._id.toHexString());
    return this.toView(updated);
  }

  async deactivateAccount(globalId: string): Promise<AccountView> {
    const account = await this.findRawByGlobalId(globalId);

    await accountCollection(this.db).updateOne(
      { _id: account._id },
      {
        $set: {
          status: 'INACTIVE',
          updatedAt: new Date(),
          version: (account.version ?? 0) + 1,
        },
      },
    );

    const updated = await this.findById(account._id.toHexString());
    return this.toView(updated);
  }

  async getByGlobalId(globalId: string): Promise<AccountView> {
    return this.toView(await this.findRawByGlobalId(globalId));
  }

  async listAccounts(first: number | null | undefined, afterCursor: string | null | undefined, statusFilter?: string | null): Promise<ConnectionView<AccountView>> {
    const size = first == null ? 20 : first;
    if (size <= 0 || size > 100) {
      throw DomainError.validation('first must be between 1 and 100');
    }

    const filter: Filter<AccountDocument> = {};

    if (statusFilter && statusFilter.trim().length > 0) {
      if (statusFilter !== 'ACTIVE' && statusFilter !== 'INACTIVE') {
        throw DomainError.validation('Invalid account status');
      }
      filter.status = statusFilter;
    }

    if (afterCursor && afterCursor.length > 0) {
      const decoded = decodeCursor(afterCursor);
      const cursorObjectId = toObjectId(decoded.id, 'Invalid cursor format');
      filter.$or = [
        { createdAt: { $lt: decoded.createdAt } },
        { createdAt: decoded.createdAt, _id: { $lt: cursorObjectId } },
      ];
    }

    const documents = await accountCollection(this.db)
      .find(filter)
      .sort({ createdAt: -1, _id: -1 })
      .limit(size + 1)
      .toArray();

    const hasNextPage = documents.length > size;
    const sliced = hasNextPage ? documents.slice(0, size) : documents;

    const edges = sliced.map((document) => ({
      cursor: encodeCursor(document.createdAt, document._id.toHexString()),
      node: this.toView(document),
    }));

    return {
      edges,
      pageInfo: {
        hasNextPage,
        hasPreviousPage: afterCursor != null,
        startCursor: edges.length > 0 ? edges[0].cursor : null,
        endCursor: edges.length > 0 ? edges[edges.length - 1].cursor : null,
      },
    };
  }

  async findRawByGlobalId(globalId: string): Promise<WithId<AccountDocument>> {
    const decoded = decodeGlobalId(globalId);
    if (decoded.type !== 'Account') {
      throw DomainError.validation('Expected Account global id');
    }
    return this.findById(decoded.id);
  }

  async findById(accountId: string): Promise<WithId<AccountDocument>> {
    const account = await accountCollection(this.db).findOne({ _id: toObjectId(accountId) });
    if (!account) {
      throw DomainError.notFound('Account not found');
    }
    return account;
  }

  async decodeAccountId(globalId: string): Promise<string> {
    const account = await this.findRawByGlobalId(globalId);
    return account._id.toHexString();
  }

  ensureAccountActive(account: WithId<AccountDocument>): void {
    if (account.status !== 'ACTIVE') {
      throw DomainError.accountInactive('Account is inactive');
    }
  }

  toView(account: WithId<AccountDocument>): AccountView {
    return {
      __typename: 'Account',
      id: encodeGlobalId('Account', account._id.toHexString()),
      ownerName: account.ownerName,
      document: account.document,
      branch: account.branch,
      number: account.number,
      currency: account.currency,
      currentBalance: decimalToString(account.currentBalance),
      status: account.status,
      version: account.version == null ? null : String(account.version),
      createdAt: account.createdAt.toISOString(),
      updatedAt: account.updatedAt.toISOString(),
    };
  }

  async getCurrentBalanceByGlobalId(accountGlobalId: string): Promise<string> {
    const account = await this.findRawByGlobalId(accountGlobalId);
    return decimalToString(account.currentBalance);
  }

  async findByIds(ids: string[]): Promise<Array<WithId<AccountDocument>>> {
    const objectIds = ids.map((id) => toObjectId(id));
    return accountCollection(this.db)
      .find({ _id: { $in: objectIds } })
      .toArray();
  }
}

function isDuplicateKeyError(error: unknown): boolean {
  if (!error || typeof error !== 'object') {
    return false;
  }

  const maybeError = error as { code?: number };
  return maybeError.code === 11000;
}

export function parseDecimalBalance(value: unknown): Decimal {
  return toDecimal(value);
}

export const accountInternals = {
  isDuplicateKeyError,
};
